Firebase/Storage/FIRStorageUploadTask.m

```diff
@@ -68,145 +70,154 @@ - (void)dealloc {
-    weakSelf.state = FIRStorageTaskStateProgress;
-    weakSelf.progress.completedUnitCount = totalBytesSent;
-    weakSelf.progress.totalUnitCount = totalBytesExpectedToSend;
-    weakSelf.metadata = self->_uploadMetadata;
-    [weakSelf fireHandlersForStatus:FIRStorageTaskStatusProgress snapshot:weakSelf.snapshot];
-    weakSelf.state = FIRStorageTaskStateRunning;
-  }];
+    NSMutableURLRequest *request = [strongSelf.baseRequest mutableCopy];
+    request.HTTPMethod = @"POST";
+    request.timeoutInterval = strongSelf.reference.storage.maxUploadRetryTime;
+    NSData *bodyData =
+        [NSData frs_dataFromJSONDictionary:[strongSelf->_uploadMetadata dictionaryRepresentation]];
+    request.HTTPBody = bodyData;
+    [request setValue:@"application/json; charset=UTF-8" forHTTPHeaderField:@"Content-Type"];
+    NSString *contentLengthString =
+        [NSString stringWithFormat:@"%zu", (unsigned long)[bodyData length]];
+    [request setValue:contentLengthString forHTTPHeaderField:@"Content-Length"];
 
-  _uploadFetcher = uploadFetcher;
+    NSURLComponents *components =
+        [NSURLComponents componentsWithURL:request.URL resolvingAgainstBaseURL:NO];
 
-  // Process fetches
-  self.state = FIRStorageTaskStateRunning;
+    if ([components.host isEqual:kGCSHost]) {
+      [components setPercentEncodedPath:[@"/upload" stringByAppendingString:components.path]];
+    }
+
+    NSDictionary *queryParams = @{@"uploadType" : @"resumable", @"name" : self.uploadMetadata.path};
+    [components setPercentEncodedQuery:[FIRStorageUtils queryStringForDictionary:queryParams]];
+    request.URL = components.URL;
+
+    GTMSessionUploadFetcher *uploadFetcher =
+        [GTMSessionUploadFetcher uploadFetcherWithRequest:request
+                                           uploadMIMEType:strongSelf->_uploadMetadata.contentType
+                                                chunkSize:kGTMSessionUploadFetcherStandardChunkSize
+                                           fetcherService:self.fetcherService];
+
+    if (strongSelf->_uploadData) {
+      [uploadFetcher setUploadData:strongSelf->_uploadData];
+      uploadFetcher.comment = @"Data UploadTask";
+    } else if (strongSelf->_fileURL) {
+      [uploadFetcher setUploadFileURL:strongSelf->_fileURL];
+      uploadFetcher.comment = @"File UploadTask";
+    }
+
+    uploadFetcher.maxRetryInterval = self.reference.storage.maxUploadRetryTime;
+
+    [uploadFetcher setSendProgressBlock:^(int64_t bytesSent, int64_t totalBytesSent,
+                                          int64_t totalBytesExpectedToSend) {
+      weakSelf.state = FIRStorageTaskStateProgress;
+      weakSelf.progress.completedUnitCount = totalBytesSent;
+      weakSelf.progress.totalUnitCount = totalBytesExpectedToSend;
+      weakSelf.metadata = self->_uploadMetadata;
+      [weakSelf fireHandlersForStatus:FIRStorageTaskStatusProgress snapshot:weakSelf.snapshot];
+      weakSelf.state = FIRStorageTaskStateRunning;
```

Comment:
Should these reference strongSelf?

Reply:
Same reason as above - I want to avoid unnecessary retain cycles. I only need the cycle in `strongSelf->_fetcherCompletion`